language/move-vm/runtime/src/native_functions.rs

```diff
@@ -1,139 +1,84 @@
+            .and_then(|funcs| funcs.get(func_name).cloned())
     }
 
-    /// Given the vector of aguments, it executes the native function.
-    pub(crate) fn dispatch(
-        self,
-        ctx: &mut impl NativeContext,
-        t: Vec<Type>,
-        v: VecDeque<Value>,
-    ) -> PartialVMResult<NativeResult> {
-        let result = match self {
-            Self::HashSha2_256 => hash::native_sha2_256(ctx, t, v),
-            Self::HashSha3_256 => hash::native_sha3_256(ctx, t, v),
-            Self::PubED25519Validate => signature::native_ed25519_publickey_validation(ctx, t, v),
-            Self::SigED25519Verify => signature::native_ed25519_signature_verification(ctx, t, v),
-            Self::VectorLength => vector::native_length(ctx, t, v),
-            Self::VectorEmpty => vector::native_empty(ctx, t, v),
-            Self::VectorBorrow => vector::native_borrow(ctx, t, v),
-            Self::VectorBorrowMut => vector::native_borrow(ctx, t, v),
-            Self::VectorPushBack => vector::native_push_back(ctx, t, v),
-            Self::VectorPopBack => vector::native_pop(ctx, t, v),
-            Self::VectorDestroyEmpty => vector::native_destroy_empty(ctx, t, v),
-            Self::VectorSwap => vector::native_swap(ctx, t, v),
-            // natives that need the full API of `NativeContext`
-            Self::AccountWriteEvent => event::native_emit_event(ctx, t, v),
-            Self::BCSToBytes => bcs::native_to_bytes(ctx, t, v),
-            Self::DebugPrint => debug::native_print(ctx, t, v),
-            Self::DebugPrintStackTrace => debug::native_print_stack_trace(ctx, t, v),
-            Self::SignerBorrowAddress => signer::native_borrow_address(ctx, t, v),
-            Self::CreateSigner => account::native_create_signer(ctx, t, v),
-            // functions below this line are deprecated and remain only for replaying old transactions
-            Self::DestroySigner => account::native_destroy_signer(ctx, t, v),
-        };
-        debug_assert!(match &result {
-            Err(e) => e.major_status().status_type() == StatusType::InvariantViolation,
-            Ok(_) => true,
-        });
-        result
+    pub fn new<I>(natives: I) -> Self
+    where
+        I: IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+    {
+        let mut map = HashMap::new();
+        for (addr, module_name, func_name, func) in natives.into_iter() {
+            let modules = get_mut_or_insert_with(&mut map, addr, || HashMap::new());
+            let funcs =
+                get_mut_or_insert_with(modules, module_name.into_string(), || HashMap::new());
+            // TODO: handle duplicates
+            funcs.insert(func_name.into_string(), func);
+        }
+        Self(map)
     }
 }
 
-pub(crate) struct FunctionContext<'a, L: LogContext> {
-    interpreter: &'a mut Interpreter<L>,
+pub struct NativeContext<'a, 'b> {
+    interpreter: &'a mut Interpreter<'b>,
```

Comment:
Here `'b` has to be a lifetime different from `'a` or otherwise this doesn't compile. With that being said I haven't bothered looking into this in depth -- will do next week.